### PR TITLE
feature(xjy): Refine PriorZero Implementation

### PR DESCRIPTION
**Pull request by xiongjyu**
这个 PR 主要完善了 PriorZero的实现与开发流程，修复了若干影响训练正确性和稳定性的关键问题，并对训练逻辑、损失计算、数据采集进行了系统性的增强。

本 PR 已完成的工作
• 修复了 PriorZero 训练流程中的多个关键 bug，包括 game segment 构建、loss 计算、log-prob 对齐以及 action 处理中的错误。
• 完善了 REINFORCE / RFT 风格的策略优化实现，在 buffer 中正确存储并使用 old_logprob，保证策略更新的正确性。
• 补充并规范了训练过程中的统计指标，包括 KL divergence、policy entropy 等，用于更好地监控训练状态。
• 优化了 Collector 与 Replay Buffer 的数据流转逻辑，提升数据一致性与采样稳定性，减少隐式错误。
• 引入并验证了单卡场景下的 vLLM 权重同步机制。
• 多 GPU / 多节点场景下的 vLLM 权重同步与稳定性验证

### CODE REVIEW

File: zoo/jericho/priorzero/priorzero_policy.py [hunk 1]
```diff
@@ -394,551 +579,100 @@ def _forward_learn(self, data: Tuple[torch.Tensor]) -> Dict[str, Union[float, in
...
-                    dtype=torch.float32
+        self._last_llm_grad_norm = 0.0
+        if self.llm_policy_cfg.enable_llm and self.llm_policy_cfg.enable_sft:
+            with self._profile_block(name="train_llm_sft"):
+                llm_sft_loss = self.compute_sft_loss(raw_obs_list=raw_obs_list, history_obs_list=history_obs_list)
+        else:
+            llm_sft_loss = torch.tensor(0.0, device=self._cfg.device)
+        if self.llm_policy_cfg.enable_llm and self.llm_policy_cfg.enable_rft:
+            with self._profile_block(name="train_llm_rft"):
+                llm_rft_loss = self.compute_rft_loss(
```
Comment (xiongjyu) [inline zoo/jericho/priorzero/priorzero_policy.py]: 这个target_value，我理解应该是和 observation 一一对应的，没有错位吧？比如target_value第一个值代表obs中第一个状态下的结果

File: zoo/jericho/priorzero/priorzero_policy.py [hunk 2]
```diff
@@ -363,20 +274,294 @@ def _init_learn(self) -> None:
...
+            token_log_probs = shifted_log_probs.gather(
+                dim=-1,
+                index=gather_labels.unsqueeze(-1)
+            ).squeeze(-1)
+            mask = (shifted_labels != -100).float()
+            token_log_probs = token_log_probs * mask
+            sequence_log_probs = token_log_probs.sum(dim=-1) / (mask.sum(dim=-1) + 1e-8)
+            
+            if self.llm_policy_cfg.rft_reward=='value':
+                rewards_tensor = torch.tensor(batch_values, device=self._cfg.device, dtype=torch.float32)
```
Comment (puyuan1996) [inline zoo/jericho/priorzero/priorzero_policy.py]: rewards_tensor 重命名为 advantage_tansor 吧？

Comment (xiongjyu) [reply zoo/jericho/priorzero/priorzero_policy.py]: ok

File: zoo/jericho/priorzero/vllm_utils/vllm_engine_ray.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    if not use_hybrid_engine:
+        # Create a big placement group to ensure that all engines are packed
+        bundles = [{"GPU": 1, "CPU": 1} for _ in range(num_engines * tensor_parallel_size)]
+        shared_pg = placement_group(bundles, strategy="PACK")
+        ray.get(shared_pg.ready())
+
+    for i in range(num_engines):
+        bundle_indices = None
+        if tensor_parallel_size > 1:
+            bundle_indices = get_bundle_indices(shared_pg, i, tensor_parallel_size)
```
Comment (puyuan1996) [inline zoo/jericho/priorzero/vllm_utils/vllm_engine_ray.py]: 这里是参考的ray官方改进吗

File: zoo/jericho/priorzero/vllm_utils/vllm_engine_ray.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                **additional_kwargs,
+            )
+        )
+
+    return vllm_engines
+
+
+def batch_vllm_engine_call(engines: List[Any], method_name: str, *args, rank_0_only: bool = True, **kwargs):
+    """
+    Batch call a method on multiple vLLM engines.
```
Comment (puyuan1996) [inline zoo/jericho/priorzero/vllm_utils/vllm_engine_ray.py]: 这里是参考的vllm的官方实现吗

File: zoo/jericho/priorzero/vllm_utils/vllm_engine_ray.py (hunk 3 above)
Comment (xiongjyu) [reply zoo/jericho/priorzero/vllm_utils/vllm_engine_ray.py]: 这个vllm_engine基本和openrlhf这部分是一样的；不过目前只使用一个vllm,并且tensor_parallel_size =1；因为显存够

File: zoo/jericho/priorzero/vllm_utils/vllm_engine_ray.py (hunk 4 above)
Comment (xiongjyu) [reply zoo/jericho/priorzero/vllm_utils/vllm_engine_ray.py]: 是的

File: zoo/jericho/priorzero/priorzero_policy.py [hunk 5]
```diff
@@ -1272,161 +537,85 @@ def _forward_collect(
...
-            policy_priors.append(policy_logits)
-
-        policy_priors = torch.stack(policy_priors).to(self._cfg.device)
-
+        for idx, actions in enumerate(valid_actions_list):
+            prior = []
+            for action in actions:
+                prior.append(llm_prior_logprob[idx][action])
+            policy_priors.append(prior)
+        policy_priors = self.pad_to_fixed_length(data=policy_priors, target_len=self.cfg.model.action_space_size, pad_val=-1e9)
```
Comment (puyuan1996) [inline zoo/jericho/priorzero/priorzero_policy.py]: 注意检查这里valid_actions_list的顺序与action_mask的对应关系

File: zoo/jericho/priorzero/priorzero_policy.py [hunk 6]
```diff
@@ -1115,60 +417,31 @@ def _monitor_vars_learn(self) -> List[str]:
         """
 
         return [
-                        # ============ LLM Loss Metrics ============
+            # ============ LLM Loss Metrics ============
             'llm_sft_loss',              # Supervised fine-tuning loss
             'llm_rft_loss',              # Reinforcement fine-tuning loss
```
Comment (puyuan1996) [inline zoo/jericho/priorzero/priorzero_policy.py]: 目前_forward_learn没有计算这些统计量吧？

File: zoo/jericho/priorzero/priorzero_collector.py [hunk 7]
```diff
@@ -91,247 +74,184 @@ def extract_raw_obs_text(obs_dict: Dict[str, Any]) -> str:
...
-                    tokenize=False,
-                    add_generation_prompt=True
+        assert self.llm_prior_generator is not None, "llm_prior_generator is None."
+        all_prompts = []
+        all_labels = []
+        for i, actions in enumerate(valid_actions_list):    
+            state = states[i]
+            history = histories[i]
+            prompt = build_llm_prompt(current_obs=state, history=history, use_cot=self.llm_policy_cfg.use_cot)
+            for action in actions:
```
Comment (puyuan1996) [inline zoo/jericho/priorzero/priorzero_collector.py]: 对于同一个state,每个action都重新构建一次prompt，这里应该可以类似下面这样优化吧：
```
def _get_llm_prior(self, states, valid_actions_list, histories):
    all_prompts, all_labels = [], []
    prompt_indices = []  # 记录每个(prompt, label)对应的prompt index
    unique_prompts = []   # 唯一的prompts

    for i, actions in enumerate(valid_actions_list):
        # 构建prompt一次
        prompt = build_llm_prompt(current_obs=states[i],
                                  history=histories[i],
                                  use_cot=False)
        prompt_idx = len(unique_prompts)
        unique_prompts.append(prompt)

        # 为每个action重用这个prompt
        for action in actions:
            all_prompts.append(prompt)  # 引用相同的字符串(不复制)
            all_labels.append(action)
            prompt_indices.append(prompt_idx)
```

File: zoo/jericho/priorzero/priorzero_llm_modules.py [hunk 8]
```diff
@@ -0,0 +1,410 @@
...
+        action_logprob_list: Optional[List[List[Any]]] = None,
+        target_values: Optional[torch.Tensor] = None,   # [B, T-1] 的 G_t
+    ) -> List[Dict[str, Any]]:
+        samples: List[Dict[str, Any]] = []
+        B = len(raw_obs_list)
+        if B == 0:
+            return samples
+        T = len(raw_obs_list[0])
+
+        for b in range(B):
```
Comment (puyuan1996) [inline zoo/jericho/priorzero/priorzero_llm_modules.py]: TODO：更高效的构建方法

File: zoo/jericho/priorzero/priorzero_entry_sync.py [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+            train_data.append(learner.train_iter)
+
+            log_vars = learner.train(train_data, collector.envstep)
+            if cfg.policy.use_priority:
+                replay_buffer.update_priority(train_data, log_vars[0]['value_priority_orig'])
+
+        if num_of_transitions >= replay_buffer.replay_buffer_size:
+            all_data = replay_buffer.sample(batch_size=replay_buffer.replay_buffer_size, policy=policy)
+            replay_buffer._clear()
+            trainer.train_rft_from_priorzero_batch(all_data)
```
Comment (puyuan1996) [inline zoo/jericho/priorzero/priorzero_entry_sync.py]: TODO：可以控制训练的off_policy程度

Comment (puyuan1996) [reply zoo/jericho/priorzero/priorzero_entry_sync.py]: TODO：llm训练速度是否过慢呢？

File: zoo/jericho/priorzero/priorzero_entry_async.py [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+        collect_kwargs = {
+            'temperature': 0.25,
+            'epsilon': 0.0
+        }
+
+        if collect_task is None or collect_task.done():
+            if coordinator.can_collect():
+                logger.info(f"\n[Iter {learner.train_iter}] Starting async collect...")
+
+                async def collect_fn():
```
Comment (puyuan1996) [inline zoo/jericho/priorzero/priorzero_entry_async.py]: collect和train异步这个目前测试通过了吗

File: zoo/jericho/priorzero/priorzero_llm_modules.py [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+
+        self._log_state(x=total_loss/gradient_accumulation_steps, name='rft_loss')
+        self.train_samples_cnt += len(samples)
+        
+        if self.vllm_engines is not None:
+            self._broadcast_to_vllm()
+        self.log_state_to_tb()
+    
+    def _broadcast_to_vllm(self):
+        use_prefix_cache = getattr(self.strategy.args, "enable_prefix_caching", False)
```
Comment (puyuan1996) [inline zoo/jericho/priorzero/priorzero_llm_modules.py]: 权重更新目前是串行的，可以改成@ray.remote的异步更新，另外应该不需要每次llm train后都更新，可以设置一个频率

File: zoo/jericho/priorzero/priorzero_entry_sync.py (hunk 9 above)
Comment (puyuan1996) [reply zoo/jericho/priorzero/priorzero_entry_sync.py]: world_model也是从replay_buffer中采样训练的吧？只清空llm训练的，不能清空world-model训练的，world_model训练是需要比较大的buffer size的

File: zoo/jericho/priorzero/priorzero_llm_modules.py [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+                
+                # optional KL(pi || ref)
+                if kl_coef > 0.0 and self.ref_model is not None:
+                    with torch.no_grad():
+                        ref_out = self.ref_model(input_ids=inputs.input_ids, attention_mask=inputs.attention_mask)
+                        ref_logits = ref_out.logits[:, :-1, :].contiguous()
+                        ref_token_logp = -F.cross_entropy(ref_logits.transpose(1, 2), shifted_labels, reduction="none")
+                        ref_token_logp = (ref_token_logp * mask)
+                        ref_seq_logp = ref_token_logp.sum(dim=-1) / (mask.sum(dim=-1) + 1e-8)
+                    kl_per_seq = compute_approx_kl(seq_logp, ref_seq_logp, kl_estimator='k2')
```
Comment (puyuan1996) [inline zoo/jericho/priorzero/priorzero_llm_modules.py]: ref_seq_logp 目前有使用吗

File: zoo/jericho/priorzero/priorzero_policy.py (hunk 6 above)
Comment (xiongjyu) [reply zoo/jericho/priorzero/priorzero_policy.py]: 这部分我修改了，现在在llm里面统计了，这个地方确实没有

File: zoo/jericho/priorzero/priorzero_entry_async.py (hunk 10 above)
Comment (xiongjyu) [reply zoo/jericho/priorzero/priorzero_entry_async.py]: 我一直都没用这个异步的；因为train不是应该等collect结束以后再sample吗？这里应该不能异步吧

File: zoo/jericho/priorzero/priorzero_llm_modules.py (hunk 11 above)
Comment (xiongjyu) [reply zoo/jericho/priorzero/priorzero_llm_modules.py]: 好的

Comment (xiongjyu) [reply zoo/jericho/priorzero/priorzero_llm_modules.py]: 但现在每次llm train完，一次要训 256 * 10条数据，应该已经很多了呀；按照batch_size=64来说的话，都更新40次模型参数了；还不需要同步吗

File: zoo/jericho/priorzero/priorzero_llm_modules.py (hunk 12 above)
Comment (xiongjyu) [reply zoo/jericho/priorzero/priorzero_llm_modules.py]: 有的

File: zoo/jericho/priorzero/priorzero_policy.py (hunk 5 above)
Comment (xiongjyu) [reply zoo/jericho/priorzero/priorzero_policy.py]: 这里检查了没问题

File: zoo/jericho/priorzero/priorzero_collector.py (hunk 7 above)
Comment (xiongjyu) [reply zoo/jericho/priorzero/priorzero_collector.py]: 这个优化好了

**Comment by puyuan1996** [comment]
We have a polished PR: https://github.com/opendilab/LightZero/pull/500
